server/svix-server/src/queue/redis.rs

```diff
@@ -114,39 +117,60 @@ async fn background_task(
+        .arg(5000);
+    // WIll be Some("OK") when set or None when not set
+    let resp: Option<String> = pool.query_async(cmd).await?;
+
+    if resp.as_deref() == Some("OK") {
+        // First look for delayed keys whose time is up and add them to the main qunue
+        let timestamp = Utc::now().timestamp();
+        let keys: Vec<String> = pool
+            .zrangebyscore_limit(&delayed_queue_name, 0isize, timestamp, 0isize, batch_size)
+            .await?;
 
-    if !keys.is_empty() {
-        // FIXME: needs to be a transaction
-        let keys: Vec<(String, String)> = pool
-            .zpopmin(&delayed_queue_name, keys.len() as isize)
-            .await
-            .unwrap();
-        let tasks: Vec<&str> = keys
-            .iter()
-            // All information is stored in the key in which the ID and JSON formated task
-            // are separated by a `|`. So, take the key, then take the part after the `|`
-            .map(|x| &x.0)
-            .map(|x| x.split('|').nth(1).expect("Improper key format"))
-            .collect();
-
-        // Then for each task, XADD them to ghe MAIN queue
-        let mut pipe = redis::pipe();
-        for task in tasks {
-            let _ = pipe.xadd(
-                &main_queue_name,
-                GENERATE_STREAM_ID,
-                &[(QUEUE_KV_KEY, task)],
-            );
+        if !keys.is_empty() {
+            // Delay every key to be scheduled ten seconds in the future so if the worker crashes
+            // during this block, the messages aren't lost
+            let mut pipe = redis::pipe();
+            for key in &keys {
+                let _ = pipe.zincr(&delayed_queue_name, key, 10_000);
+            }
+            let _: () = pool.query_async_pipeline(pipe).await?;
+
+            let tasks: Vec<&str> = keys
+                .iter()
+                // All information is stored in the key in which the ID and JSON formated task
+                // are separated by a `|`. So, take the key, then take the part after the `|`
+                .map(|x| x.split('|').nth(1).expect("Improper key format"))
+                .collect();
+
+            // Then for each task, XADD them to ghe MAIN queue
+            let mut pipe = redis::pipe();
+            for task in tasks {
+                let _ = pipe.xadd(
+                    &main_queue_name,
+                    GENERATE_STREAM_ID,
+                    &[(QUEUE_KV_KEY, task)],
+                );
+            }
+            let _: () = pool.query_async_pipeline(pipe).await?;
+
+            // Finally remove the tasks from the delayed queue so they aren't resent
+            let _: () = pool
+                .query_async(Cmd::zrem(&delayed_queue_name, keys))
+                .await?;
         }
-        let _: () = pool.query_async_pipeline(pipe).await?;
-    } else {
-        // Wait for half a second before attempting to fetch again if nothing was found
-        sleep(Duration::from_millis(500)).await;
```

Comment:
Things are very messy with the two unrelated tasks happening in the same function (both the xpending and the zrange). I think what we should do is split them to two separate functions and honestly just loop over them separately with different sleep schedules, etc. This is just very confusing and it's easy to put in mistakes (like you have with moving this sleep).

Reply:
See https://github.com/svix/svix-webhooks/pull/573